pulser-core/pulser/operators/operator.py

```diff
@@ -0,0 +1,419 @@
+        """Defines the sum of two OperatorString."""
+        if not isinstance(operator, OperatorString):
+            raise TypeError("Right operand for + must be an OperatorString")
+        return OperatorString(
+            self.coefficients + operator.coefficients,
+            self.operators + operator.operators,
+        )
+
+    def __rmul__(self, scalar: Number) -> OperatorString:
+        """Multiplies the current OperatorString by a number."""
+        if not isinstance(scalar, Number):
+            raise TypeError("Left operand with * must be a Number.")
+        return OperatorString(
+            [coeff * scalar for coeff in self.coefficients], self.operators
+        )
+
+    def kron(
+        self,
+        operator: OperatorString,
+        qubit_ids1: Collection[QubitId],
+        qubit_ids2: Collection[QubitId],
+    ) -> OperatorString:
+        """Kronecker product of the current operator with a new operator."""
+        if not set(self.qubit_ids).intersection(operator.qubit_ids):
+            raise ValueError("Operators should have different qubit ids.")
+        coeffs = []
+        ops = []
+        for coeff1, global_op1 in zip(self.coefficients, self.operators):
+            inner_op1 = [global_op1]
+            if "global" in global_op1.targeted_qubits:
+                assert len(global_op1.kron_operations) == 1
+                inner_op1 = [
+                    TargetedOperatorString(
+                        ((global_op1.kron_operations[0][0], set([qubit_id])),)
+                    )
+                    for qubit_id in qubit_ids1
+                ]
+            for op1 in inner_op1:
+                for coeff2, global_op2 in zip(
+                    operator.coefficients, operator.operators
+                ):
+                    inner_op2 = [global_op2]
+                    if "global" in global_op2.targeted_qubits:
+                        assert len(global_op2.kron_operations) == 1
+                        inner_op2 = [
+                            TargetedOperatorString(
+                                (
+                                    (
+                                        global_op2.kron_operations[0][0],
+                                        set([qubit_id]),
+                                    ),
+                                )
+                            )
+                            for qubit_id in qubit_ids2
+                        ]
+                    for op2 in inner_op2:
+                        coeffs.append(coeff1 * coeff2)
+                        ops.append(op1.kron(op2))
+        return OperatorString(coeffs, ops)
+
+
+class TimeOperatorString(UserList[OperatorString]):
+
+    def __init__(self, operators: list[OperatorString]):
+        super().__init__(operators)
+
+    def operations(self):
+        return tuple(self.data)
+
+
+@dataclass
+class Operator(ABC):
+    """Defines a generic operator class."""
+
+    operator_string: OperatorString
```

Comment:
Should this class not have a `from_string` abstract method or something, that forces backends to specify how to use the `operator_string` to actually create an instance? That way you can do all required operator releated things through this class.

Reply:
So in this implementation I have investigated a class that could be common to all backends, that all backends could inherit from and just implement a to_myobject() and a from_myobject() method - see QutipOperator in pulser-simulation/operators.
The advantage of this approach seemed at first that some checks could be shared with all the backends, but the issue is that to perform any operation you should:
- either implement your own algebra on Operator (solution I favored)
- convert the Operator into your objects [to_myobject()], make the operation with these objects, convert the resulting object into an Operator [from_myobject()]. This seems to be very costly, see my comment in pulser_simulation/operators.

Reply:
And so in the current implementation, the from_string is the __init__ of the Operator class. I thought of this implementation when writing QutipOperator. I realized then that converting an object into a string is complex and takes time, whereas if you store the string you have it immediately when you want to send your operator to another backend. That's why I favoured this implementation rather than storing the object and converting the object into an OperatorString when needed (some operations might take more time (I don't think that addition is so long compared to addition in qutip) but the main operation being the conversion from object to OperatorString, this will take less time) 

If you

Reply:
If you still favor storing the object and having an abstract class method from_operator_string (and an abstract method to_operator_string), I am still open to this, just let me know if my arguments made sense :)

Reply:
For this reason, i would recommend a minimal implementation like so:
```
class Operator(ABC):
    @abstractmethod
    def __add__(self, operator: Operator) -> Operator:
        pass
     @abstractmethod
    def __rmul__(self, scalar: Number) -> Operator:
        pass
    @abstractmethod
    def __mul__(self, state: State) -> State:
        pass
    @abstractmethod
    def __matmul__(self, operator: Operator) -> Operator:
        pass
    @abstractmethod
    def kron(self, operator: Operator) -> Operator:
        pass
    @abstractmethod
    def from_string(self, string:OperatorString) -> Operator:
        pass
    @abstractmethod
    def to_string(self) -> OperatorString:
        pass
```
this implies we need some `State(ABC)` similar to how we have the `OperatorString` and the `Operator`. If we want to implement the algebra on `OperatorString` we could have it inherit from `Operator`. Note that i removed the

Reply:
I guess we could just have `OperatorString` inherit from `Operator` anyway, since it implements some of the methods, and for the others we could throw an exception

Reply:
Thanks for your feedback @a-quelle. I realized yesterday that I was indeed not using ABC correctly, and I now fully understand the way to go.